app/multitenant/consul_pipe_router.go

```diff
@@ -0,0 +1,605 @@
+}
+
+// Atomically modify a pipe in a callback.
+// If pipe doesn't exist you'll get nil in callback.
+func (pr *consulPipeRouter) get(key string) (*consulPipe, error) {
+	var (
+		kv   = pr.client.KV()
+		pipe consulPipe
+	)
+	kvp, _, err := kv.Get(key, queryOptions)
+	if err != nil {
+		return nil, err
+	}
+	if kvp == nil {
+		return nil, nil
+	}
+	if err := pipe.fromBytes(kvp.Value); err != nil {
+		return nil, err
+	}
+	return &pipe, nil
+}
+
+// Atomically modify a pipe in a callback.
+// If pipe doesn't exist you'll get nil in callback.
+func (pr *consulPipeRouter) cas(key string, f func(*consulPipe) (*consulPipe, bool, error)) (*consulPipe, error) {
+	var (
+		index   = uint64(0)
+		kv      = pr.client.KV()
+		pipe    *consulPipe
+		retries = 10
+		retry   = true
+	)
+	for i := 0; i < retries; i++ {
+		kvp, _, err := kv.Get(key, queryOptions)
+		if err != nil {
+			log.Errorf("Error getting %s: %v", key, err)
+			continue
+		}
+		if kvp != nil {
+			pipe = &consulPipe{}
+			if err := pipe.fromBytes(kvp.Value); err != nil {
+				log.Errorf("Error deserialising pipe %s: %v", key, err)
+				continue
+			}
+			index = kvp.ModifyIndex // if it doesn't exist, it will be 0
+		}
+
+		if pipe, retry, err = f(pipe); err != nil {
+			log.Errorf("Error CASing pipe %s: %v", key, err)
+			if !retry {
+				return nil, err
+			}
+			continue
+		}
+
+		if pipe == nil {
+			panic("Callback must instantiate pipe!")
+		}
+
+		value, err := pipe.toBytes()
+		if err != nil {
+			log.Errorf("Error serialising pipe %s: %v", key, err)
+			continue
+		}
+		ok, _, err := kv.CAS(&consul.KVPair{
+			Key:         key,
+			Value:       value,
+			ModifyIndex: index,
+		}, writeOptions)
+		if err != nil {
+			log.Errorf("Error CASing pipe %s: %v", key, err)
+			continue
+		}
+		if !ok {
+			log.Errorf("Error CASing pipe %s, trying again %d", key, index)
+			continue
+		}
+		return pipe, nil
+	}
+	return nil, fmt.Errorf("Failed to aquire pipe")
+}
+
+// Watch a given pipe and trigger a callback when it changes.
+// if callback returns false or error, exit (with the error).
+func (pr *consulPipeRouter) watch(key string, deadline time.Time, f func(*consulPipe) (bool, error)) (*consulPipe, error) {
+	var (
+		index = uint64(0)
+		kv    = pr.client.KV()
+	)
+	for deadline.After(mtime.Now()) {
+		// Poll waiting for the entry to get updated
+		kvp, meta, err := kv.Get(key, &consul.QueryOptions{
+			RequireConsistent: true,
+			WaitIndex:         index,
+			WaitTime:          longPollDuration,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Error getting %s: %v", key, err)
+		}
+		if kvp == nil {
+			return nil, fmt.Errorf("Pipe %s unexpectedly deleted!", key)
+		}
+
+		pipe := &consulPipe{}
+		pipe.fromBytes(kvp.Value)
+		if ok, err := f(pipe); !ok {
+			return pipe, nil
+		} else if err != nil {
+			return pipe, err
+		}
+
+		index = meta.LastIndex
+	}
+	return nil, fmt.Errorf("Timed out waiting on %s", key)
```

Comment:
looping and calling `Get` could miss changes while running the callback. Doesn't consul offer a `subscribe`?